File: ticktick_cli/cli/tag.py
"""Tag commands for TickTick CLI."""

from typing import Annotated

import typer

from ticktick_cli import output
from ticktick_cli.cli.errors import handle_client_error
from ticktick_cli.client import get_client

app = typer.Typer(help="Manage tags")


@app.command("list")
@handle_client_error
def list_tags(
    json_output: Annotated[
        bool,
        typer.Option("--json", help="Output as JSON"),
    ] = False,
) -> None:
    """List all tags."""
    client = get_client()
    tags = client.get_tags()

    if json_output:
        output.print_json(tags)
    elif not tags:
        output.print_info("No tags found")
    else:
        output.print_tags_table(tags)


@app.command("add")
@handle_client_error
def add_tag(
    name: Annotated[str, typer.Argument(help="Tag name")],
    color: Annotated[
        str | None,
        typer.Option("--color", "-c", help="Hex color code"),
    ] = None,
    parent: Annotated[
        str | None,
        typer.Option("--parent", "-p", help="Parent tag name (for hierarchical tags)"),
    ] = None,
    json_output: Annotated[
        bool,
        typer.Option("--json", help="Output as JSON"),
    ] = False,
) -> None:
    """Create a new tag."""
    client = get_client()

    kwargs = {"name": name}
    if color:
        kwargs["color"] = color
    if parent:
        kwargs["parent"] = parent

    tag = client.create_tag(**kwargs)

    if json_output:
        output.print_json(tag)
    else:
        output.print_success(f"Created tag: {name}")


@app.command("rename")
@handle_client_error
def rename_tag(
    old_name: Annotated[str, typer.Argument(help="Current tag name")],
    new_name: Annotated[str, typer.Argument(help="New tag name")],
) -> None:
    """Rename a tag."""
    client = get_client()
    client.rename_tag(old_name, new_name)
    output.print_success(f"Renamed '{old_name}' to '{new_name}'")


@app.command("edit")
@handle_client_error
def edit_tag(
    name: Annotated[str, typer.Argument(help="Tag name")],
    color: Annotated[
        str | None,
        typer.Option("--color", "-c", help="New hex color code"),
    ] = None,
    json_output: Annotated[
        bool,
        typer.Option("--json", help="Output as JSON"),
    ] = False,
) -> None:
    """Edit an existing tag."""
    client = get_client()

    kwargs = {}
    if color:
        kwargs["color"] = color

    if not kwargs:
        output.print_error("No fields to update")
        raise typer.Exit(1)

    tag = client.update_tag(name, **kwargs)

    if json_output:
        output.print_json(tag)
    else:
        output.print_success(f"Updated tag: {name}")


@app.command("delete")
@handle_client_error
def delete_tag(
    names: Annotated[list[str], typer.Argument(help="Tag name(s) to delete")],
    force: Annotated[
        bool,
        typer.Option("--force", "-f", help="Skip confirmation"),
    ] = False,
) -> None:
    """Delete tag(s)."""
    if not force:
        confirm = typer.confirm(f"Delete {len(names)} tag(s)?")
        if not confirm:
            output.print_info("Cancelled")
            return

    client = get_client()
    for name in names:
        client.delete_tag(name)
        output.print_success(f"Deleted: {name}")

File: src/commands/tag.ts
/**
 * Tag commands.
 */
import { Command } from "commander";
import { getClient } from "../api/client.js";
import {
  printError,
  printSuccess,
  printInfo,
  printJson,
  printTagsTable,
} from "../output/index.js";
import { handleError } from "./errors.js";
import { getGlobalOptions } from "../index.js";

export function createTagCommand(): Command {
  const tag = new Command("tag").description("Manage tags");

  // list command
  tag
    .command("list")
    .description("List all tags")
    .option("--json", "Output as JSON")
    .action(
      handleError(async function (this: Command, options) {
        const globalOpts = getGlobalOptions(this);
        const client = await getClient({ validation: globalOpts.validation });
        const tags = await client.getTags();

        if (options.json) {
          printJson(tags);
        } else if (tags.length === 0) {
          printInfo("No tags found");
        } else {
          printTagsTable(tags);
        }
      })
    );

  // add command
  tag
    .command("add <name>")
    .description("Create a new tag")
    .option("-c, --color <hex>", "Tag color (hex code)")
    .option("-p, --parent <name>", "Parent tag name")
    .option("--json", "Output as JSON")
    .action(
      handleError(async function (this: Command, name: string, options) {
        const globalOpts = getGlobalOptions(this);
        const client = await getClient({ validation: globalOpts.validation });

        const tagData: Parameters<typeof client.createTag>[0] = { name };

        if (options.color) {
          tagData.color = options.color;
        }
        if (options.parent) {
          tagData.parent = options.parent;
        }

        const createdTag = await client.createTag(tagData);

        if (options.json) {
          printJson(createdTag);
        } else {
          printSuccess(`Created tag: ${createdTag.name}`);
        }
      })
    );

  // rename command
  tag
    .command("rename <oldName> <newName>")
    .description("Rename a tag")
    .action(
      handleError(async function (this: Command, oldName: string, newName: string) {
        const globalOpts = getGlobalOptions(this);
        const client = await getClient({ validation: globalOpts.validation });

        // Verify tag exists
        const tags = await client.getTags();
        const foundTag = tags.find((t) => t.name === oldName);

        if (!foundTag) {
          printError(`Tag not found: ${oldName}`);
          process.exit(1);
        }

        await client.renameTag(oldName, newName);
        printSuccess(`Renamed tag: ${oldName} → ${newName}`);
      })
    );

  // edit command
  tag
    .command("edit <name>")
    .description("Edit an existing tag")
    .option("-c, --color <hex>", "New color (hex code)")
    .option("-p, --parent <name>", "New parent tag name")
    .option("--json", "Output as JSON")
    .action(
      handleError(async function (this: Command, name: string, options) {
        const globalOpts = getGlobalOptions(this);
        const client = await getClient({ validation: globalOpts.validation });

        // Find the tag first
        const tags = await client.getTags();
        const foundTag = tags.find((t) => t.name === name);

        if (!foundTag) {
          printError(`Tag not found: ${name}`);
          process.exit(1);
        }

        const updateData: Parameters<typeof client.updateTag>[0] = { name };

        if (options.color) {
          updateData.color = options.color;
        }
        if (options.parent) {
          updateData.parent = options.parent;
        }

        const updatedTag = await client.updateTag(updateData);

        if (options.json) {
          printJson(updatedTag);
        } else {
          printSuccess(`Updated tag: ${name}`);
        }
      })
    );

  // delete command
  tag
    .command("delete <name>")
    .description("Delete a tag")
    .option("-f, --force", "Skip confirmation")
    .action(
      handleError(async function (this: Command, name: string) {
        const globalOpts = getGlobalOptions(this);
        const client = await getClient({ validation: globalOpts.validation });

        // Verify tag exists
        const tags = await client.getTags();
        const foundTag = tags.find((t) => t.name === name);

        if (!foundTag) {
          printError(`Tag not found: ${name}`);
          process.exit(1);
        }

        await client.deleteTag(name);
        printSuccess(`Deleted tag: ${name}`);
      })
    );

  return tag;
}
